// react-day-picker/lib/types/UIElement.js
/**
 * Represent an element or an element state of the DayPicker user interface,
 * whose style or class name can be changed using the `styles` or `classNames`
 * prop.
 *
 * ```
 * <DayPicker
 *   classNames={{ captionLast: 'my-custom-class' }}
 *   styles={{ captionLast: { color: 'red' }}}
 * />
 * ```
 */
export var UIElement;
(function (UIElement) {
    /** The container element. */
    UIElement["Root"] = "root";
    /** The container element when number of months is greater than 1. */
    UIElement["RootMultipleMonths"] = "rootMultipleMonths";
    /** The caption element (showing the calendar heading and the navigation) */
    UIElement["Caption"] = "caption";
    /** The caption element when the first of multiple months */
    UIElement["CaptionFirst"] = "captionFirst";
    /** The caption element when the last of multiple months */
    UIElement["CaptionLast"] = "captionLast";
    /** The caption element is between */
    UIElement["CaptionBetween"] = "captionBetween";
    /** The container of the drop-downs. */
    UIElement["CaptionDropdowns"] = "dropdownsContainer";
    /** The drop-down select element. */
    UIElement["Dropdown"] = "dropdown";
    /** The container of the caption’s label (e.g. "September 2021"). */
    UIElement["CaptionLabel"] = "dropdownLabel";
    /** The drop-down to change the month. */
    UIElement["DropdownMonth"] = "dropdownMonth";
    /** The drop-down to change the year. */
    UIElement["DropdownYear"] = "dropdownYear";
    /** The icon aside the drop-down. */
    UIElement["DropdownIcon"] = "dropdownIcon";
    /** The wrapper of the months (there may be more months visualized when `numberOfMonths`). */
    UIElement["Months"] = "months";
    /** The wrapper of the table displaying the month. */
    UIElement["Month"] = "month";
    /** The table displaying the calendar. */
    UIElement["Table"] = "table";
    /** The body element of the table displaying the calendar. */
    UIElement["TBody"] = "body";
    /** The head element of the table displaying the calendar. */
    UIElement["Head"] = "head";
    /** The row element of the head. */
    UIElement["HeadRow"] = "headRow";
    /** The cell element of the head. */
    UIElement["HeadCell"] = "headCell";
    /** The navigation element. */
    UIElement["Nav"] = "nav";
    /** A navigation button. */
    UIElement["NavButton"] = "navButton";
    /** The "previous month" navigation button. */
    UIElement["NavButtonPrev"] = "navButtonPrev";
    /** The "next month" navigation button. */
    UIElement["NavButtonNext"] = "navButtonNext";
    /** The icon contained in navigation button. */
    UIElement["NavIcon"] = "navIcon";
    /** A row in the table element (each row shows a week). */
    UIElement["Row"] = "row";
    /** A row head in the table element, used to display the week numbers. */
    UIElement["RowHead"] = "rowHead";
    /** The element containing the week number */
    UIElement["WeekNumber"] = "weekNumber";
    /** The cell containing the day element. */
    UIElement["Cell"] = "cell";
    /** The day element (a button when the day is interactive). */
    UIElement["Day"] = "day";
})(UIElement || (UIElement = {}));
